src/wagtail_vector_index/index/models.py

```diff
@@ -0,0 +1,402 @@
+
+    def _get_querysets(self) -> Sequence[models.QuerySet]:
+        return self.querysets
+
+    def get_converter(self) -> EmbeddableFieldsDocumentConverter:
+        queryset_models = [qs.model for qs in self._get_querysets()]
+        all_the_same = len(set(queryset_models)) == 1
+        if not all_the_same:
+            raise ValueError(
+                "All querysets must be of the same model to use the default converter."
+            )
+        return self.converter_class(queryset_models[0])
+
+    def rebuild_index(self):
+        """Before building an index, generate Embedding objects for everything in
+        this index"""
+        querysets = self._get_querysets()
+
+        for queryset in querysets:
+            for instance in queryset:
+                self.get_converter().generate_embeddings(
+                    instance, embedding_backend=self.embedding_backend
+                )
+
+        return super().rebuild_index()
+
+    def get_documents(self) -> Iterable[Document]:
+        querysets = self._get_querysets()
+
+        for queryset in querysets:
+            instances = queryset.prefetch_related("embeddings")
+            yield from self.get_converter().bulk_to_documents(
+                instances, embedding_backend=self.embedding_backend
+            )
+
+
+class PageVectorIndex(ModelVectorIndex):
+    """A model vector indexed for use with Wagtail pages that automatically
+    restricts indexed models to live pages."""
+
+    querysets: Sequence[PageQuerySet]
+
+    def _get_querysets(self) -> list[PageQuerySet]:
+        qs_list = super()._get_querysets()
+
+        # Technically a manager instance, not a queryset, but we want to use the custom
+        # methods.
+        return [cast(PageQuerySet, qs).live() for qs in qs_list]
+
+
+# ###########
+# Classes related to automatic generation of indexes for models
+# ###########
+
+
+class GeneratedIndexMixin(models.Model):
+    """Mixin for Django models that automatically generates registers a VectorIndex for the model"""
+
+    vector_index_class: ClassVar[Optional[type[VectorIndex]]] = None
+    converter_class: ClassVar[type[DocumentConverter]]
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def get_vector_index(cls):
+        """Get a vector index instance for this model"""
+
+        # If the user has specified a custom `vector_index_class`, use that
+        if cls.vector_index_class:
+            index_cls = cls.vector_index_class
+        # If the model is a Wagtail Page, use a special PageVectorIndex
+        elif issubclass(cls, Page):
+            index_cls = PageVectorIndex
+        # Otherwise use the standard ModelVectorIndex
+        else:
+            index_cls = ModelVectorIndex
+
+        return type(
+            f"{cls.__name__}Index",
+            (index_cls,),
+            {
+                "querysets": [cls.objects.all()],
+                "converter_class": cls.converter_class,
+            },
+        )()
```

Comment:
**Thoughts:** In some ways, it feels weird we create an index instance here, rather than use the registry, don't you think? Wouldn't we expect the user to fetch it from the registry once it's been already created?

In some ways `get` does not mean `instantiate` to me. I think they should be separate methods.

Reply:
This is a good thought, I've had a go at refactoring this so `GeneratedIndexMixin` has a `build_vector_index_class` classmethod that actually builds the class - this is called by `register_indexed_models` on app `ready`.

The convenient accessor for the index has been moved to a `classproperty` so now you can just use `MyModel.vector_index`.

Let me know what you think,

Reply:
I think that's fine. I don't know if there's an implication of keeping a reference to it on a model... If I delete it from a registry, it'll stay on the model. Wouldn't it make sense to start with just a getter?

Reply:
There's no reference to it on the model, it still fetches it from the registry, so if you're manually messing with the registry for some reason (we don't document a delete behaviour there), then it will error as expected

Reply:
Ahhh, it's a decorator! I did not know it was possible nice!